llo/plugin.go

```diff
@@ -0,0 +1,924 @@
+	if outctx.SeqNr <= 1 {
+		if len(ao.Observation) != 0 {
+			return fmt.Errorf("Observation is not empty")
+		}
+	}
+
+	var observation Observation
+	// FIXME: do we really want to allow empty observations? happens because "" is not valid JSON
+	if len(ao.Observation) > 0 {
+		err := json.Unmarshal(ao.Observation, &observation)
+		if err != nil {
+			return fmt.Errorf("Observation is invalid json (got: %q): %w", ao.Observation, err)
+		}
+	}
+
+	if p.PredecessorConfigDigest == nil && len(observation.AttestedPredecessorRetirement) != 0 {
+		return fmt.Errorf("AttestedPredecessorRetirement is not empty even though this instance has no predecessor")
+	}
+
+	if len(observation.AddChannelDefinitions) > MAX_OBSERVATION_ADD_CHANNEL_DEFINITIONS_LENGTH {
+		return fmt.Errorf("AddChannelDefinitions is too long: %v vs %v", len(observation.AddChannelDefinitions), MAX_OBSERVATION_ADD_CHANNEL_DEFINITIONS_LENGTH)
+	}
+
+	if len(observation.RemoveChannelIDs) > MAX_OBSERVATION_REMOVE_CHANNEL_IDS_LENGTH {
+		return fmt.Errorf("RemoveChannelIDs is too long: %v vs %v", len(observation.RemoveChannelIDs), MAX_OBSERVATION_REMOVE_CHANNEL_IDS_LENGTH)
+	}
+
+	if len(observation.StreamValues) > MAX_OBSERVATION_STREAM_VALUES_LENGTH {
+		return fmt.Errorf("StreamValues is too long: %v vs %v", len(observation.StreamValues), MAX_OBSERVATION_STREAM_VALUES_LENGTH)
+	}
+
+	for streamID, obsResult := range observation.StreamValues {
+		if obsResult.Valid && obsResult.Val == nil {
+			return fmt.Errorf("stream with id %q was marked valid but carries nil value", streamID)
+		}
+	}
+
+	return nil
+}
+
+type Outcome struct {
+	// LifeCycleStage the protocol is in
+	LifeCycleStage commontypes.LLOLifeCycleStage
+	// ObservationsTimestampNanoseconds is the median timestamp from the
+	// latest set of observations
+	ObservationsTimestampNanoseconds int64
+	// ChannelDefinitions defines the set & structure of channels for which we
+	// generate reports
+	ChannelDefinitions commontypes.ChannelDefinitions
+	// Latest ValidAfterSeconds value for each channel, reports for each channel
+	// span from ValidAfterSeconds to ObservationTimestampSeconds
+	ValidAfterSeconds map[commontypes.ChannelID]uint32
+	// StreamMedians is the median observed value for each stream
+	// QUESTION: Can we use arbitrary types here to allow for other types or
+	// consensus methods?
+	StreamMedians map[commontypes.StreamID]*big.Int
+}
+
+// The Outcome's ObservationsTimestamp rounded down to seconds precision
+func (out *Outcome) ObservationsTimestampSeconds() (uint32, error) {
+	result := time.Unix(0, out.ObservationsTimestampNanoseconds).Unix()
+	if int64(uint32(result)) != result {
+		return 0, fmt.Errorf("timestamp doesn't fit into uint32: %v", result)
+	}
+	return uint32(result), nil
+}
+
+// Indicates whether a report can be generated for the given channel.
+// Returns nil if channel is reportable
+func (out *Outcome) IsReportable(channelID commontypes.ChannelID) error {
```

Comment:
why is this at the channel level and not the stream level? 

I see this: 
```
	for _, streamID := range channelDefinition.StreamIDs {
		if out.StreamMedians[streamID] == nil {
			return errors.New("IsReportable=false; median was nil")
		}
	}
```
does that mean if a single stream is failing - the entire channel fails to report?

Reply:
Hmm, good question. It think it probably does need to be at the channel level because otherwise what would be the outcome?

Imagine you try to create a report and one of the streams is bad. The datatype has no way to mark this. Do you just write zeroes? That could lead to some really unexpected and bad results. I think we cannot bake a report at all if any one of the streams is bad.